Update calling-c-and-fortran docs (#31700)
JuliaLang/julia: doc/src/manual/calling-c-and-fortran-code.md

```

The macro [`@cfunction`](@ref) generates the C-compatible function pointer for a call to a
Julia function. The arguments to [`@cfunction`](@ref) are:

1. A Julia Function

> 2. The return type

Author: That depends on how you interpret it.  If you think about the return type as belonging to the function, then I think it should be `the function's return type` (which I shortened to `the return type` here), since the function was introduced above.  If it's an arbitrary return type, then `a return type` would be better.

I'll change it to `the function's return type`, since that's clearer.

> 2. The return type

Reviewer: OK, I guess I was reading it as "A call's return type"

3. A literal tuple of input types

As with `ccall`, all of these arguments will be evaluated at compile-time when the containing method is defined.

Currently, only the platform-default C calling convention is supported. This means that
`@cfunction`-generated pointers cannot be used in calls where WINAPI expects `stdcall`
function on 32-bit windows, but can be used on WIN64 (where `stdcall` is unified with the
C calling convention).

> As a classic example, consider the standard C library `qsort` function, declared as:

Author: Reverted.

ccall((:foo, "libfoo"), Cvoid, (Int32, Float64), x, y)
```

will behave as if the following were written[^3]:

```julia
ccall((:foo, "libfoo"), Cvoid, (Int32, Float64),

>       Base.convert(Int32, x), Base.convert(Float64, y))

Reviewer 2: Why change this and remove the text below? This is not how it works and gives the wrong impression.

Author: The text below is preserved as a footnote.  I wrote it this way because I felt that, for most users, `cconvert` followed by `unsafe_convert` was more detail than they would want to know, and that this gets the point across (i.e., that the passed values are converted to the appropriate type).  That said, I'm willing to change it back if you don't think the footnote is enough or appropriate.

Reviewer 2: I mean, this is the docs for how to `ccall` so why should I have to scroll to a footnote in order to see how `ccall` works?

Reviewer 3: >  for most users, cconvert followed by unsafe_convert was more detail than they would want to know, and that this gets the point across (i.e., that the passed values are converted to the appropriate type).

For any user, they should either not care about anything about conversion for `ccall` arguments at all, or if they do, they must not related it to `convert`. This change gets the wrong point across. I've recently seen people still thinking `convert` is the method to use for this and it's just wrong.

Bottom line is that I don't really mind if this is said in the footnote or not, but you must not mention `convert`.

Packed structs and union declarations are not supported by Julia.

You can get an approximation of a `union` if you know, a priori, the field that will have
the greatest size (potentially including padding). When translating your fields to Julia, declare
the Julia field to be only of that type.

> Arrays of parameters can be expressed with `NTuple`.  For example, this struct in C,

Reviewer: As written, it seems to need a colon, but that's suboptimal because it doesn't match the second clause, which has a comma. Fixing the sentence to have parallel grammar structure resolves the  punctuation issue.

```

Here, the input `p` is declared to be of type `Ref{gsl_permutation}`, meaning that the memory

> that `p` points to may be managed by Julia or by C. A pointer to memory allocated by C should generally

Reviewer: ```suggestion
that `p` points to may be managed by Julia or by C. A pointer to memory allocated by C should
```

> be of type `Ptr{gsl_permutation}`, but it is convertible using [`Base.cconvert`](@ref) and therefore

Reviewer: ```suggestion
be of type `Ptr{gsl_permutation}`. But `Ptr` is a subtype of `Ref`, and therefore `Ref` be used in any covariant (argument) context.
The `Ref` supertype will accept a `Ptr`, and treats it the same as if it had been restricted to just accept `Ptr`.
Additionally, this makes it possible to pass a handle to memory that has been allocated by Julia. 
```

Reviewer: no, you can never use `pointer`. the rest sounds ok

Author: Okay, would I just pass `a` in that second example then?

> can be used in the same (covariant) context of the input argument to a [`ccall`](@ref).

Reviewer: ```suggestion
```

> If the C wrapper never expects the user to pass pointers to memory managed by Julia, then using

Reviewer: ```suggestion
Now if you look closely enough at this example, you may notice that it is incorrect, given our explanation above of preferred declaration types. Do you see it? The function we are calling is going to free the memory. This type of operation cannot be given a Julia object (it will crash or cause memory corruption). Therefore, it may be preferable to declare the `p` type as `Ptr{gsl_permutation }`, to make it harder for the user to mistakenly pass another sort of object there than one obtained via `gsl_permutation_alloc`.

If the C wrapper never expects the user to pass pointers to memory managed by Julia, then using
```

```

The C function wrapped returns an integer error code; the results of the actual evaluation of
the Bessel J function populate the Julia array `result_array`. This variable is declared as a
`Ref{Cdouble}`, since its memory is allocated and managed by Julia. The implicit call to 
[`Base.cconvert(Ref{Cdouble}, result_array)`](@ref) unpacks
the Julia pointer to a Julia array data structure into a form understandable by C.

> 

Reviewer: 👍

## Fortran Wrapper Example

> The following example computes a dot product using a Fortran BLAS function.  Note the use

Reviewer: ```suggestion
The following example utilizes ccall to call a function in a common Fortran library (libBLAS) to computes a dot product. Notice that the argument mapping is a bit different here than above, as we need to map from Julia to Fortran.
```

Author: Here too: will clean up line length later.

> of `Ref` and `Ptr`.

Reviewer: ```suggestion
On every argument type, we specify `Ref` or `Ptr`. This mangling convention may be specific to your fortran compiler and operating system, and is likely undocumented. However, wrapping each in a `Ref` (or `Ptr`, where equivalent) is a frequent requirement of Fortran compiler implementations:
```

For direct C++ interfacing, see the [Cxx](https://github.com/Keno/Cxx.jl) package. For tools to create C++
bindings, see the [CxxWrap](https://github.com/JuliaInterop/CxxWrap.jl) package.

[^1]: Non-library function calls in both C and Julia can be inlined and thus may have

>     even less overhead than calls to shared library functions. When both libraries and executables are

Reviewer: ```suggestion
    even less overhead than calls to shared library functions.
    The point above is that the cost of actually doing foreign function call is about the same as doing a call in either native language.
```

>     generated by LLVM, it is possible to perform whole-program optimizations that can even optimize

Reviewer: ```suggestion
```

Reviewer 4: just out of curiosity, is this impossible for now?

Reviewer: no, some packages do it, it's just largely nonsense and irrelevant for the manual

>     across this boundary, but Julia does not yet support that. In the future, however, it may do so,

Reviewer: ```suggestion
```

>     yielding even greater performance gains.

Reviewer: ```suggestion
```
